src/Analyzer/Passes/FuseFunctionsPass.cpp

```diff
@@ -0,0 +1,135 @@
+#include <Analyzer/Passes/FuseFunctionsPass.h>
+
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <Functions/FunctionFactory.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/ConstantNode.h>
+
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypeTuple.h>
+
+namespace DB
+{
+
+namespace
+{
+
+class FuseFunctionsMatcher : public InDepthQueryTreeVisitor<FuseFunctionsMatcher>
+{
+public:
+
+    static bool matchFunctionName(const String & name)
+    {
+        return name == "count" || name == "sum" || name == "avg";
+    }
+
+    void visitImpl(QueryTreeNodePtr & node)
+    {
+        auto * function_node = node->as<FunctionNode>();
+        if (!function_node || !function_node->isAggregateFunction() || !matchFunctionName(function_node->getFunctionName()))
+            return;
+
+        auto argument_hash = function_node->getArgumentsNode()->getTreeHash();
+        mapping[argument_hash].push_back(&node);
+    }
+
+    struct QueryTreeHashForMap
+    {
+        size_t operator()(const IQueryTreeNode::Hash & hash) const { return hash.first ^ hash.second; }
+    };
+
+    /// argument -> list of sum/count/avg functions with this argument
+    std::unordered_map<IQueryTreeNode::Hash, std::vector<QueryTreeNodePtr *>, QueryTreeHashForMap> mapping;
+};
+
+template <typename... Args>
+QueryTreeNodePtr createResolvedFunction(ContextPtr context, const String & name, DataTypePtr result_type, Args &&... args)
```

Comment:
I would prefer to not make this function template just array or QueryTreeNodes as last argument.